### PR TITLE
Sync with play master

### PR DESCRIPTION
**Pull request by godenji**
Since `getOptional[T]` was removed from play master in [this commit](https://github.com/playframework/playframework/pull/4588), anyone who builds play from source and depends on the play mailer plugin (without pulling in the play 2.4.2 dep) will have their application terminate on each and every sent email due to `NoSuchMethodError` on `getOptional[T]` o_O

Making matters worse, Akka swallows the actual cause (NSME) leaving one with a joyful stack trace [like this](http://pastebin.com/ECSP5V0P) which is about as helpful as, "error: good luck".

Anyway, this is very much an edge case, most play-ers are unaffected, but it would be nice to provide snapshot builds of plugins in future so those of us that build from source can avoid above situation.

### CODE REVIEW

**Comment by godenji** [comment]
Looks like Sonatype has the missing deps but play version is hashed a la "2.5.0-2015-08-17-e0f466a-SNAPSHOT" instead of the "2.5.0-SNAPSHOT" referenced in this PR.

Not sure how to workaround...at least the PR is documented for anyone else that may run into this show stopper in future.

**Comment by Mogztter** [comment]
@godenji I don't think we need to build against the latest SNAPSHOT (at least for now). Could you please update your pull request to remove newline at the end of files and revert the play version to 2.4.2 ?

**Comment by godenji** [comment]
@Mogztter sure, 2.4.2 would be the sensible approach...if it actually worked ;-)

`MailerPlugin.scala:297: could not find implicit value for parameter loader: play.api.ConfigLoader[Option[String]]`

What's at issue here is that play master provides [the magic](https://github.com/richdougherty/playframework/commit/cf60c65b551cedf99a5cceaee5b15cc9cff8a37b#diff-4378e6503263d81d7869f6056b31e6b1R1092) for `get[Option[T]]` to work.

Basically play mailer has to depend on play master.

**Comment by Mogztter** [comment]
Ah! I thought that `get[Option[T]]` was compatible with Play 2.4.2 :disappointed: 

> Looks like Sonatype has the missing deps but play version is hashed a la "2.5.0-2015-08-17-e0f466a-SNAPSHOT" instead of the "2.5.0-SNAPSHOT" referenced in this PR.

Yes, I don't know how to reference the latest `2.5.0-SNASPHOT`... @jroper must know ?

**Comment by godenji** [comment]
@Mogztter yes, a chicken and egg problem, which is why mailer master depends on play 2.4.2 and not latest snapshot.

It'd be great to provide snapshot builds of the mailer plugin (all play plugins really) so those of us living on the edge with play master can be the guinea pigs that we are while knowing that anything that breaks is related to *current* code and not the previous version ;-)

Anyway, maybe an sbt-ism exists to search for latest snapshot in a repo.

**Comment by Mogztter** [comment]
Maybe you could try `2.5.0+` ?
http://www.scala-sbt.org/0.13/tutorial/Library-Dependencies.html#Ivy+revisions
https://ant.apache.org/ivy/history/2.3.0/ivyfile/dependency.html#revision

**Comment by Mogztter** [comment]
https://github.com/sbt/sbt-unique-version#how-to-point-at-it

**Comment by godenji** [comment]
@Mogztter does Travis rebuild automatically here? Couldn't find sbt 0.13.8 which must be a spurious build error.

**Comment by Mogztter** [comment]
Yes, I can manually trigger builds from Travis :)

**Comment by Mogztter** [comment]
I just have one concern with this pull request. I don't know if `master` should build against the latest Play! version because we have a shorter release cycle than Play! core.
New features are merged into `master` but if `master` depends on a SNAPSHOT, we will have to wait until the next major Play! version is released to release a stable version of the mailer plugin (with all the new features). Sure we can backport every single features but that's extra work.

Another issue is that we will break semantic versioning. Version 3.x builds against Play 2.4.x and we will need to create a new 4.x version that will build against Play 2.5.x.
If I want to introduce incompatible API changes within the Play 2.4.x series I can't do it because version 4.x is already taken.

Wdyt ?

**Comment by godenji** [comment]
@Mogztter agreed, tricky.

Maintenance-wise it's probably easier for maintainers to continue with the status quo. The drawback is that without this PR the plugin is unusable (and potentially dangerous) for users building against play master.

Saying that, how about a play mailer 2.5.x branch? Until play 2.5 is released that could be the staging area for integrating latest play master changes into the plugin. Not sure how important semantic versioning is in this case, could be an arbitrary version name since it will always build against latest snapshot version of play master.

**Comment by Mogztter** [comment]
> Saying that, how about a play mailer 2.5.x branch? Until play 2.5 is released that could be the staging area for integrating latest play master changes into the plugin. Not sure how important semantic versioning is in this case, could be an arbitrary version name since it will always build against latest snapshot version of play master.

Yes I think this is probably the best option. Maybe we can use the latest version from the "stable" branch and add a suffix like `-next` ? Ex:

 * `master` branch builds against the latest stable version from Play! core (version `3.0.2-SNAPSHOT`)
 * `next` branch builds against the latest SNASPHOT from Play! core (version `3.0.2-next-SNASPHOT`)

That way the next branch doesn't take away version number. We could also remove the version number on the `next` branch but I'm not sure we will be able to publish this "version" on Maven repositories.

**Comment by Mogztter** [comment]
@godenji In order for me to merge your work, could you please point to the new `next` branch ?

**Comment by godenji** [comment]
@Mogztter sorry for the delay, massively busy all of a sudden.

**Comment by Mogztter** [comment]
> sorry for the delay, massively busy all of a sudden.

No worries :smile: 

I saw that you changed the version but could you also edit your pull request on GitHub to target `playframework:next` branch instead of `playframework:master` because your code need to be merged in the `next` branch not the `master`.

Thanks,
Guillaume.

**Comment by Mogztter** [comment]
@godenji If this is unclear or if you don't have time I can also manually cherry-pick your fix on the next branch and close this pull request ?

**Comment by godenji** [comment]
>  I can also manually cherry-pick your fix on the next branch and close this pull request

@Mogztter yes, please do, swamped here.

Thanks

**Comment by Mogztter** [comment]
Merged into next branch (commits cherry-picked)

https://github.com/playframework/play-mailer/commit/149925c33f01d2eab4448edb976cfcb54d644c50
https://github.com/playframework/play-mailer/commit/b0ef21402a59d397dac5b50babe7d92f78e8e5a9

**Comment by godenji** [comment]
@Mogztter thanks!
